## Intro to Data Visualization with Python

## Part 1 

###  Step 1.  📟 Importing libraries and dataset

### Step 2. Data Wrangling

### Step 3. Combining dataframes

### Step 4. Importing data set

###  Step 1.  📟 Importing libraries and dataset

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Tell Python to remember a main folder path
path = r'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [3]:
path

'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [4]:
# importing dataframe
df = pd.read_csv(os.path.join(path, 'Original Data', 'customers.csv'))

### Step 2. Data Wrangling

In [5]:
df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df.shape

(206209, 10)

In [7]:
# renaming columns
df = df.rename(columns={ 
    'Surnam' : 'Last Name', 
    'STATE' : 'State', 
    'date_joined' : 'Date Joined', 
    'n_dependants' : 'Number of Dependents', 
    'fam_status' : 'Family Status', 
    'income' : 'Income'})

In [8]:
df = df.rename(columns={'User ID' : 'user_id'})

In [9]:
df.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependents Family Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [10]:
df.describe()

user_id            Age  Number of Dependents         Income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [11]:
# finding missing values
missing_values = df.isnull().sum()

In [12]:
missing_values

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Date Joined                 0
Number of Dependents        0
Family Status               0
Income                      0
dtype: int64

In [17]:
# df['First Name'].fillna('Unknown', inplace=True)


In [19]:
# most_frequent_name = df['First Name'].mode()[0]
df['First Name'].fillna(most_frequent_name, inplace=True)


### Comments: 
We can see that there are **11259 missing values** in the first name column which can be either fillied with the most frequent name or replace them with unknown. 
I would like to leave them as it is for the moment. 

In [20]:
missing_values

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Date Joined                 0
Number of Dependents        0
Family Status               0
Income                      0
dtype: int64

In [13]:
# finding duplicate rows
duplicates = df.duplicated()

In [15]:
# summing up the total number of duplicates
num_duplicates = duplicates.sum()

In [16]:
num_duplicates

0

### Comments: 
There no duplicates found at the moment in the data frame. 

In [21]:
# finding column types
df.dtypes

user_id                  int64
First Name              object
Last Name               object
Gender                  object
State                   object
Age                      int64
Date Joined             object
Number of Dependents     int64
Family Status           object
Income                   int64
dtype: object

## Comments: 
No such unusual column to be found in data types.

### Step 3. Combining dataframes

In [22]:
# importing dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge_grouped.pkl'))

In [23]:
# combining customer with ords_prods_merge
merged_df = pd.merge(ords_prods_merge, df, on='user_id', how='inner')

In [24]:
# checking merge
merged_df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   ...   order_frequency_flag First Name  Last Name  Gender    State Age  \
0  ...  Non-frequent customer      Linda     Nguyen  Female  Alabama  31   
1  ...  Non-frequent customer      Linda     Nguyen  Female  Alabama  31   
2  ...  Non-frequent customer      Linda     Nguyen  Female  Alabama  31   
3  ...  Non-frequent customer      Linda     Nguyen  Female  Alabama  31   
4  ...  Non-frequent customer      Linda     Nguyen  Female  Alabama  31   

  Date Joined  Number of Dependents Family Status  Income  
0   2/17/2019                     3       married   40423  
1   2/17/2019                     3       married   40423  
2   2/17/2019                     3       married   40423  
3   2/17/2019                     3       married   40423  
4   2/17/2019                     3       married   40423  

[5 rows x 32 columns]

In [25]:
merged_df.shape

(32435059, 32)

### Step 4. Importing data set

In [26]:
#exporting as a pickle
merged_df.to_pickle(os.path.join(path,'Prepared Data', 'ords_prods_customer_merge.pkl'))